### PR TITLE
Collapse identical first buckets

### PR DESCRIPTION
**Pull request by kennyzzhang**
Here's a forgotten optimization, should help a bit

### CODE REVIEW

**Review by Victor-C-Zhang — COMMENTED** [review]
Can you explain this more?

**Comment by kennyzzhang** [comment]
> 
> 
> Can you explain this more?

We have columns of buckets, within each column we have varying guesses for the number of nonzero elements.
The first bucket of each column contains all indices, the second bucket contains half of all indices, the third contains a quarter, and so on.
Since every column's first bucket contains all indices, they are all identical.
We can collapse them into a single bucket and tack this bucket onto the end of our array of buckets.

File: include/bucket.h [hunk 1]
```diff
@@ -31,6 +31,16 @@ namespace Bucket_Boruvka {
...
+  /**
+   * Checks whether a Bucket is good, assuming the Bucket contains all elements.
+   * @param a The bucket's a value.
+   * @param c The bucket's c value.
+   * @param n Size of the vector being sketched.
+   * @param sketch_seed The seed of the Sketch this Bucket belongs to.
+   * @return true if this Bucket is good, else false.
+   */
+  inline static bool is_good(const vec_t& a, const vec_hash_t& c, const vec_t& n, const long& sketch_seed);
+
```
Comment (Victor-C-Zhang) [inline include/bucket.h:43]: for readability put these after the other `contains` function

File: include/sketch.h [hunk 2]
```diff
@@ -10,7 +10,7 @@
 #include <gtest/gtest_prod.h>
 
 #define bucket_gen(x, c) double_to_ull((c)*(log2(x)+1))
-#define guess_gen(x) double_to_ull(log2(x)+2)
+#define guess_gen(x) double_to_ull(log2(x)+1)
```
Comment (Victor-C-Zhang) [inline include/sketch.h:13]: have you tested this rigorously?

Comment (Victor-C-Zhang) [reply include/sketch.h:13]: actually nvm this should be fine

File: l0_sampling/sketch.cpp [hunk 3]
```diff
@@ -19,11 +19,12 @@ void Sketch::update(const vec_t& update_idx) {
   const unsigned num_buckets = bucket_gen(n, num_bucket_factor);
   const unsigned num_guesses = guess_gen(n);
   XXH64_hash_t update_hash = Bucket_Boruvka::index_hash(update_idx, seed);
+  Bucket_Boruvka::update(bucket_a[num_buckets * num_guesses], bucket_c[num_buckets * num_guesses], update_idx, update_hash);
   for (unsigned i = 0; i < num_buckets; ++i) {
     col_hash_t col_index_hash = Bucket_Boruvka::col_index_hash(i, update_idx, seed);
     for (unsigned j = 0; j < num_guesses; ++j) {
       unsigned bucket_id = i * num_guesses + j;
-      if (Bucket_Boruvka::contains(col_index_hash, 1 << j)){
+      if (Bucket_Boruvka::contains(col_index_hash, 2 << j)){
```
Comment (Victor-C-Zhang) [inline l0_sampling/sketch.cpp:27]: 1 << (j+1)

File: l0_sampling/sketch.cpp [hunk 4]
```diff
@@ -44,13 +45,20 @@ vec_t Sketch::query() {
...
+
   for (unsigned i = 0; i < num_buckets; ++i) {
     for (unsigned j = 0; j < num_guesses; ++j) {
       unsigned bucket_id = i * num_guesses + j;
-      if (bucket_a[bucket_id] != 0 || bucket_c[bucket_id] != 0) {
+      if (all_buckets_zero && (bucket_a[bucket_id] != 0 || bucket_c[bucket_id] != 0)) {
         all_buckets_zero = false;
       }
-      if (Bucket_Boruvka::is_good(bucket_a[bucket_id], bucket_c[bucket_id], n, i, 1 << j, seed)) {
+      if (Bucket_Boruvka::is_good(bucket_a[bucket_id], bucket_c[bucket_id], n, i, 2 << j, seed)) {
```
Comment (Victor-C-Zhang) [inline l0_sampling/sketch.cpp:61]: 1 << (j+1)

File: l0_sampling/sketch.cpp [hunk 5]
```diff
@@ -85,16 +93,24 @@ bool operator== (const Sketch &sketch1, const Sketch &sketch2) {
...
+     << "a:" << sketch.bucket_a[num_buckets * num_guesses] << std::endl
+     << "c:" << sketch.bucket_c[num_buckets * num_guesses] << std::endl
+     << (Bucket_Boruvka::is_good(sketch.bucket_a[num_buckets * num_guesses], sketch.bucket_c[num_buckets * num_guesses], sketch.n, sketch.seed) ? "good" : "bad") << std::endl;
+
   for (unsigned i = 0; i < num_buckets; ++i) {
     for (unsigned j = 0; j < num_guesses; ++j) {
       unsigned bucket_id = i * num_guesses + j;
       for (unsigned k = 0; k < sketch.n; k++) {
-        os << (Bucket_Boruvka::contains(Bucket_Boruvka::col_index_hash(i, k, sketch.seed), 1 << j) ? '1' : '0');
+        os << (Bucket_Boruvka::contains(Bucket_Boruvka::col_index_hash(i, k, sketch.seed), 2 << j) ? '1' : '0');
```
Comment (Victor-C-Zhang) [inline l0_sampling/sketch.cpp:108]: same

**Comment by Victor-C-Zhang** [comment]
If there's documentation about sketching, update it. Especially with column hashing/etc this is getting very complicated and unintuitive.

**Comment by etwest** [comment]
Hey Kenny and Victor, this is something we probably want to get into the final system for running our experiments. What still needs to be done before this is ready to be merged?

**Comment by kennyzzhang** [comment]
I think it's functional code wise, but there's some code style/documentation things that we could change because what we're doing is kinda unintuitive.

**Comment by etwest** [comment]
> I think it's functional code wise, but there's some code style/documentation things that we could change because what we're doing is kinda unintuitive.

Currently we don't really have documentation in this repo. Maybe just add a note to the README which indicates what the change was so that we can explain it elsewhere.

**Comment by Victor-C-Zhang** [comment]
> Hey Kenny and Victor, this is something we probably want to get into the final system for running our experiments. What still needs to be done before this is ready to be merged?

Once the commented parts are fixed, I can approve.

**Review by etwest — APPROVED** [review]
I'm just going to approve this because I want it for the final system. Maybe we'll come back to it later.

**Pull request merged**